tests/votinghistory.py

```diff
@@ -0,0 +1,164 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+import sys
+import os
+import io
+
+import e2e_args
+import getpass
+import time
+import logging
+import multiprocessing
+from random import seed
+import infra.ccf
+import infra.proc
+import infra.jsonrpc
+import infra.remote
+import json
+import ledger
+import msgpack
+
+import cryptography.x509
+from cryptography.hazmat.primitives.asymmetric import ec
+from cryptography.hazmat.backends import default_backend
+
+
+def verify_sig(raw_cert, sig, req):
+    cert = cryptography.x509.load_der_x509_certificate(
+        raw_cert, backend=default_backend()
+    )
+    pub_key = cert.public_key()
+    hash_alg = ec.ECDSA(cert.signature_hash_algorithm)
+    pub_key.verify(sig, req, hash_alg)
+
+
+def run(args):
+    hosts = ["localhost", "localhost"]
+
+    ledger_filename = None
+
+    with infra.ccf.network(
+        hosts, args.build_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
+    ) as network:
+        primary, others = network.start_and_join(args)
+
+        # create a lua query file to change a member state to accepted
+        with open("query.lua", "w") as qfile:
+            qfile.write(
+                """local tables, param = ...
+               local member_id = param
+               local STATE_ACCEPTED = 0
+               local member_info = {status = STATE_ACCEPTED}
+               local p = Puts:new()
+               p:put("members", member_id, member_info)
+               return Calls:call("raw_puts", p)"""
+            )
+
+        # create json file to be passed as the argument to the query.lua file
+        # it is passing a member id
+        with open("param.json", "w") as pfile:
+            pfile.write("""{"p": 0}""")
+
+        # propose to add a new member
+        # proposal number 0
+        infra.proc.ccall("./genesisgenerator", "cert", "--name=member4")
+        result = infra.proc.ccall(
```

Comment:
I wonder if we should extend the `ccf` module to include native python client code for the member RPCs as they are rather verbose?

Reply:
Sounds reasonable, but it would not be in the scope of this PR.